### PR TITLE
Fix JAnsi bug in Gradle

### PR DESCRIPTION
**Pull request by lacasseio**
This PR fix https://github.com/fusesource/jansi/issues/69 directly in Gradle by duplicating the required code and applying our patch to the code. This also fix https://github.com/gradle/gradle/issues/882 and related issues.

Note the `console.flush()` was moved down to prevent event replay if an exception is thrown during the flush.

### CODE REVIEW

**Comment by BruceEckel** [comment]
Is there a way to use this from my gradle-wrapper.properties file? Specifically, what would I replace:
distributionUrl=https\://services.gradle.org/distributions/gradle-3.3-bin.zip
with? Thanks.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/sink/AnsiConsoleUtil.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+
+import static org.fusesource.jansi.internal.CLibrary.STDOUT_FILENO;
+import static org.fusesource.jansi.internal.CLibrary.isatty;
+import static org.fusesource.jansi.internal.Kernel32.*;
+
+/**
+ * @see <a href="https://github.com/fusesource/jansi/issues/69">fusesource/jansi#69</a>
```
Comment (big-guy) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/sink/AnsiConsoleUtil.java]: Can you provide a link to the commit this was copied from?

File: subprojects/logging/src/main/java/org/gradle/internal/logging/sink/AnsiConsoleUtil.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        protected void processCursorRight(int count) throws IOException {
+            getConsoleInfo();
+            info.cursorPosition.x = (short) Math.min(info.window.width(), info.cursorPosition.x + count);
+            applyCursorPosition();
+        }
+
+        @Override
+        protected void processCursorDown(int count) throws IOException {
+            getConsoleInfo();
+            info.cursorPosition.y = (short) Math.min(Math.max(0, info.size.y - 1), info.cursorPosition.y + count);
```
Comment (big-guy) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/sink/AnsiConsoleUtil.java]: Can you put a comment here so future us knows what we added?

**Comment by big-guy** [comment]
@BruceEckel This would need to be merged and then built into a nightly, which are [available here](https://gradle.org/gradle-nightly-build/).

File: subprojects/logging/src/main/java/org/gradle/internal/logging/sink/AnsiConsoleUtil.java (hunk 2 above)
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/sink/AnsiConsoleUtil.java]: Done. Added a reference to this PR: https://github.com/fusesource/jansi/pull/70

File: subprojects/logging/src/main/java/org/gradle/internal/logging/sink/AnsiConsoleUtil.java (hunk 1 above)
Comment (lacasseio) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/sink/AnsiConsoleUtil.java]: I added comments that links to the right commit as well as highlight the copied code.

**Comment by big-guy** [comment]
LGTM, can you paste a link to a nightly that contains this change once it's merged?

**Comment by lacasseio** [comment]
@big-guy This nightly contains the change: https://services.gradle.org/distributions-snapshots/gradle-3.4-20170116000026+0000-bin.zip

**Pull request merged**